src/pubsub.js

```diff
@@ -1,162 +1,128 @@
-      emitSelf: false,
-      gossipIncoming: true,
-      fallbackToFloodsub: true,
-      ...options
-    }
-  }
+    this.seenCache = new TimeCache()
 
-  /**
-   * When a peer has dialed into another peer, it sends its subscriptions to it.
-   * @override
-   * @param {PeerInfo} peerInfo The peer dialed
-   * @param {Connection} conn  The connection with the peer
-   * @param {Function} callback
-   *
-   * @returns {void}
-   */
-  _onDial (peerInfo, conn, callback) {
-    const idB58Str = peerInfo.id.toB58String()
-
-    super._onDial(peerInfo, conn, (err) => {
-      if (err) return callback(err)
+    /**
+     * Pubsub options
+     */
+    this._options = _options
 
-      const peer = this.peers.get(idB58Str)
-      if (peer && peer.isWritable) {
-        // Immediately send my own subscription to the newly established conn
-        peer.sendSubscriptions(this.subscriptions)
-      }
-      nextTick(() => callback())
-    })
+    this._onRpc = this._onRpc.bind(this)
   }
 
   /**
-   * Dial a received peer.
+   * Peer connected successfully with pubsub protocol.
    * @override
-   * @param {PeerInfo} peerInfo The peer being dialed
-   * @param {function} callback
-   *
-   * @returns {void}
+   * @param {PeerInfo} peerInfo peer info
+   * @param {Connection} conn connection to the peer
+   * @returns {Promise<void>}
    */
-  _dialPeer (peerInfo, callback) {
-    callback = callback || function noop () { }
+  async _onPeerConnected (peerInfo, conn) {
+    await super._onPeerConnected(peerInfo, conn)
     const idB58Str = peerInfo.id.toB58String()
-
-    // If already have a PubSub conn, ignore
     const peer = this.peers.get(idB58Str)
-    if (peer && peer.isConnected) {
-      return nextTick(() => callback())
-    }
-
-    // If already dialing this peer, ignore
-    if (this._dials.has(idB58Str)) {
-      this.log('already dialing %s, ignoring dial attempt', idB58Str)
-      return nextTick(() => callback())
-    }
-
-    // Verify if is known that the peer does not support Gossipsub
-    const onlySupportsFloodsub = peerInfo.protocols.has(floodsubMulticodec) && !peerInfo.protocols.has(this.multicodec)
-
-    // Define multicodec to use
-    // Should fallback to floodsub if fallback is enabled, protocols were negotiated, and no Gossipsub available
-    let multicodec = this.multicodec
 
-    if (this._options.fallbackToFloodsub && onlySupportsFloodsub) {
-      multicodec = floodsubMulticodec
+    if (peer && peer.isWritable) {
+      // Immediately send my own subscriptions to the newly established conn
+      peer.sendSubscriptions(this.subscriptions)
     }
-
-    this._dials.add(idB58Str)
-    this.log('dialing %s %s', multicodec, idB58Str)
-
-    this.libp2p.dialProtocol(peerInfo, multicodec, (err, conn) => {
-      this.log('dial to %s complete', idB58Str)
-      this._dials.delete(idB58Str)
-
-      if (err) {
-        // If previously dialed gossipsub and not supported, try floodsub if enabled fallback
```

Comment:
Is the floodsub fallback gone in this refactor?

Reply:
No! Basically, all the dial logic that existed in this repo, as well as in `floodsub`, `dht` got extracted to an upper layer, as it was mostly replicated.

In this cases, we pass up an array of multicodecs, and it will attempt to dial both `gossipsub` and `floodsub` and call the `onConnect` handler